use borgstore and other big changes

ThomasWaldmann opened the pull request:
Fixes #8330 (no cache sync anymore).

Fixes #8325 (no segments, replaying segments anymore).

Fixes #7377, fixes #7379 (no transactions, no refcounting anymore).

Fixes #7154 (new repo locking code).

Fixes #6983 (no repo index any more).

Fixes #6899 (no compact segments any more).

Fixes #6121, fixes #7278 (no cache sync any more, no archives fetching).

Fixes #6567 (not needed any more).

Fixes #6331 (hints file not used anymore).

Fixes #6291, fixes #6289 (no segment files any more, no DEL tags).

Fixes #6288 in a different way (borg rspace command to reserve disk space that can be freed in case of emergencies).

Fixes #5654, fixes #6057, fixes #6094, fixes #7154 (new lock implementation for borgstore).

Fixes #5514 (new locking system allows shared locks for most).

Fixes #5261 (bypass-lock was removed).

Fixes #5050 (no hints any more).

Fixes #4827 (no persistent chunks cache, no cache sync anymore).

Fixes #4438 (not possible, delete/prune just kill the root reference now and don't look at anything else).

Fixes #4428 (separate objects in config/*).

Fixes #4004 - most borg commands now can use a shared repository lock (exceptions: borg check and compact).

Fixes #3128. Fixes #3196. (no cache sync anymore)

Fixes #2454, fixes #2398, fixes #3036 (no commits, no transactions any more, no log-like/append-only segments, new check implementation).

Fixes #2681, fixes #2571 (no cache sync, no chunks.archive.d).

Fixes #2454 (no commits, no transactions anymore).

Fixes #2444 (new borg check implementation, no LoggedIO code used any more).

Fixes #1293 (solved in a different way, delete/prune are super fast now).

Fixes #1244 (no transactions anymore).

Fixes #916, fixes #474, fixes #1766 (no LocalCache any more, no cache transaction any more).

Fixes #768 - most borg commands now can use a shared repository lock (exceptions: borg check and compact).

Maybe builds the foundation to solve / work on:
- #102 (S3 and other clouds, without FUSE)
- #1070 (sftp: already works)
- #5324


## new repository based on borgstore project

stores chunks into separate files (not: segment/pack files, at least for now).

borgstore has a very simple api that makes implementing backends easy.

in borg 2.0, this will primarily use the "file:" backend from borgstore to implement `file:` and `ssh:` repositories, but long term we might go away from the borg.remote code (RPC api via ssh) and just use a "remote" borgstore.

there is also a `sftp:` repository now implemented via the respective borgstore backend. more might be coming, even cloud stuff should be easily possible with that (PRs welcome!).

## repository: convergence rather than transactions

borg 1.x used the segment files in a log-like way (only appending new stuff at the end) and implemented transactions via a COMMIT tag - if the transaction was not completed (no COMMIT at the end), it rolled back the incomplete transaction to the last commit.

the code implementing transactions was rather complex and required an exclusive lock on the repo for correct operations.

borg2 now just adds repo objects in the right order, first pushing referenced objects, then the references to them. even if an operation is interrupted, nothing bad happens.
there might be some unreferenced objects for a while, but they will get referenced if the operation is retried later and completes. borg compact will deal with anything not needed.

## no checkpoint archives, no .borg_part files anymore

saving them was only needed due to the transactional/rollback behavior of borg1.

borg2 does not do that rollback any more, so the checkpoints are not needed.
the user can just re-run the interrupted command and it will notice that some stuff is already present in the repo and only transfer new stuff.

## new borg compact doing garbage collection

borg compact is still needed to free space in the repo, but it doesn't really need to "compact segments" as there are no segment files anymore. so it will do less I/O to move stuff around in the repo.

but maybe some sort of segment/pack files will come back later, so we'll just keep the command name.

borg compact is now doing more work that was previously done by borg delete and borg check: it determines which chunks are not used anymore (and removes them). because it needs to read the archives for that, borg compact now needs the borg key.

## super fast borg delete and borg prune

as borg does not do precise refcounting any more, delete now just kills the archive from `archives/*` (removing the reference to its root) and lets borg compact clean up all the now unreferenced chunks.

## new repo locking code

repository locking code using borgstore locks/*.

locks auto-expire and get deleted if they don't get refreshed regularly (this is good to clean up stale locks of now-dead remote borg processes).

locks also get deleted if their owner process is dead (this is good to clean up stale locks of now-dead local borg processes).

most borg commands now can use a shared repository lock (exceptions: borg check and compact, which must use an exclusive lock).

## new repo config / repo key storage

Config is now stored into separate files config/* - less risk (e.g. for the repokey) if other config items need updating.

Repokey now stored into keys/* (only 1 key for now).

## new manifest storage

the manifest chunk that also had the list of archives inside was split into config/manifest and separate files archives/*.


some features might come back later (stats, quota, append-only, ...)

ThomasWaldmann commented:
mypy still unhappy: src/borg/testsuite/locking3.py:5: error: Skipping analyzing "borgstore.store": module is installed, but missing library stubs or py.typed marker  [import-untyped]

ThomasWaldmann commented on src/borg/helpers/misc.py:
```diff
@@ -2,7 +2,7 @@
 import os
 import os.path
-import platform
+import platform  # python stdlib import - if this fails, check that pwd != src/borg/
```
cwd not pwd

ThomasWaldmann commented on src/borg/testsuite/archiver/transfer_cmd.py:
```diff
@@ -35,6 +35,7 @@ def check_repo():
     check_repo()
 
     archiver.repository_location = original_location + "2"
+    # XXX broken: --other-repo only accepts "Repository", but not "Repository3" repos:
```
TODO?

ThomasWaldmann commented on src/borg/archiver/__init__.py:
```diff
@@ -36,7 +36,7 @@
     from ..helpers import msgpack
     from ..helpers import sig_int
-    from ..remote import RemoteRepository
+    from ..remote import RemoteRepository  # TODO 3?
```
TODO?

ThomasWaldmann commented:
Guess this needs some review. Anybody?

m3nu commented:
I'm about 25% through and this looks like a great change to decouple storage logic and simplify more things. Still want to run this branch locally to see how it works. Anyways, solid job @ThomasWaldmann!

ThomasWaldmann commented:
@buengese I suspect it could be quite similar to the rclone REST server for restic, complemented by a REST client/backend in borgstore.

m3nu commented:
Did some more testing of this PR and didn't encounter any errors. Tested those common cases:

- Create local repo, add some archives and files, extract single file
- Create remote repo over SSH, add some archives and files, extract single file
- Resume interrupted backup to remote repo

Anything you would want to see tested more specifically, @ThomasWaldmann ?

jdchristensen commented:
I'd be interested in seeing some timing comparisons between this and the previous repo format for various operations under different workloads, including with larger repos with many archives, and also with both local and remote repos.  (I'd volunteer to try this myself, but I'm swamped right now.)

ThomasWaldmann commented:
@m3nu if you have time:

- try big backups, scalability?
- try sftp://user@host:port/abs/olute/path/to/repo with a big backup, scalability?
- performance comparison with borg 1.x?
- try accessing a repo in parallel (currently, repo locking should behave similar as in borg 1.x, but is a new implementation)

ThomasWaldmann commented:
@m3nu quota, storage use info, append-only are all not yet implemented.

when running `borg compact -v`, you get some infos, but even that can be further improved.

ThomasWaldmann commented:
`borg compact -v` now gives some more and more useful infos:
- repo size
- compression and dedup factor
- how much space it freed (if any)

ThomasWaldmann commented:
@m3nu Yes. The idea is "if you can do it at the source, you can also do it at the repo". At least, due to dedup, it doesn't go up proportional to the archives count. Due to non-deduped files and due to long files being split into multiple chunks, the repo files count will be higher than the src files count.

Of course this has pros and cons, but that was also the case for segment files and segment compaction.

When doing performance comparisons, we have to consider that some bookkeeping work shifted its place and some stuff works rather differently now.

E.g. borg1 delete/prune could take rather long, because it was precisely counting down the refcounts and writing DEL tags to the repo when they reached zero. borg2 (this PR) now just kills the archive root reference from the archives list and leaves the rest of the work to borg compact (which has to determine "is a chunk still needed" for all chunks anyway).

borg1 check also used to do this (finding "orphan chunks"), but as borg2 compact does this now, I have removed it from borg2 check. borg2 compact also notices if chunks needed for healing have appeared and then advises to run borg check --repair. same if borg2 compact notices that you have lost chunks that are in use by archives.

and, of course, borg2 compact does NOT copy around huge amounts of data any more to free space as borg1 did with the segment files to remove their "holes".

borg2 create could be a bit faster due to not checkpointing any more (which IIRC also saved the indexes/caches), which borg1 did every 30 mins by default and more often if people modified it for unstable connections. for borg2, it behaves like "always checkpointing" (because there are no repo transaction rollbacks any more), but there is no overhead at all. 

OTOH, borg2 create could also be a bit slower, because more files are opened/closed and it dynamically loads the existing chunk ids list from the repo at every invocation.

so there likely is quite some optimization potential still, but I consider this way of doing it also as a door-opener for the future, with more parallelism / less locking and more repo backend options. so even if borg2 create takes longer, it could get overall faster in the future, because we hopefully will be able to run multiple backups to same repo at the same time.

m3nu commented:
Right. It's a trade-off at multiple levels. I believe Restic has a related [setting](https://restic.readthedocs.io/en/latest/047_tuning_backup_parameters.html#pack-size), pack size, to control if one wants more or fewer data segments. To get around limitations in the number of files on the backend.

ThomasWaldmann commented:
guess I am mostly done here for now.

some stuff needs server-side support (e.g. append-only and enforcing quotas by borg), so that should be delayed until that is decided/reimplemented (e.g. by a borg rest server / borgstore rest client).

ThomasWaldmann commented:
I did another pass through the issue tracker and added "fixes xxxx" (meaning: fixed in master branch, after this gets merged) to the topmost comment, also roughly sorted these.

We'll get rid of a lot of issues, but due to quite some new code, maybe also gain some new ones.

ThomasWaldmann commented:
Hmm, I either need some feedback to do more work here or this should get merged ASAP due to the widespread changes causing conflicts for other PRs.

ThomasWaldmann commented:
TODO: encrypt `archives/*` objects (contain archive name, ID, timestamp).

witten commented:
Initial feedback... Running the following on a non-existent repo directory (which borgmatic does to probe whether a configured repo already exists prior to `rcreate`):

```bash
$ ~/Downloads/borg-env/bin/borg rinfo --json --repo ~/Downloads/test.borg 
```

Results in this (exit code 2):

```
Local Exception

Error:

BackendDoesNotExist: posixfs storage base path does not exist or is not a directory: /home/witten/Downloads/test.borg

If reporting bugs, please include the following:

Traceback (most recent call last):
  File "/home/witten/Downloads/borg/src/borg/archiver/__init__.py", line 622, in main
    exit_code = archiver.run(args)
                ^^^^^^^^^^^^^^^^^^
  File "/home/witten/Downloads/borg/src/borg/archiver/__init__.py", line 516, in run
    rc = func(args)
         ^^^^^^^^^^
  File "/home/witten/Downloads/borg/src/borg/archiver/_common.py", line 151, in wrapper
    with repository:
  File "/home/witten/Downloads/borg/src/borg/repository.py", line 133, in __enter__
    self.open(exclusive=bool(self.exclusive), lock_wait=self.lock_wait, lock=self.do_lock)
  File "/home/witten/Downloads/borg/src/borg/repository.py", line 180, in open
    self.store.open()
  File "/home/witten/Downloads/borg-env/lib/python3.12/site-packages/borgstore/store.py", line 65, in open
    self.backend.open()
  File "/home/witten/Downloads/borg-env/lib/python3.12/site-packages/borgstore/backends/posixfs.py", line 54, in open
    raise BackendDoesNotExist(
borgstore.backends.errors.BackendDoesNotExist: posixfs storage base path does not exist or is not a directory: /home/witten/Downloads/test.borg

Platform: Linux flux 6.1.99-1-MANJARO #1 SMP PREEMPT_DYNAMIC Mon Jul 15 21:35:02 UTC 2024 x86_64
Linux: Unknown Linux  
Borg: 2.0.0b10.dev80+g74a67749  Python: CPython 3.12.4 msgpack: 1.0.8 fuse: None [pyfuse3,llfuse]
PID: 2588631  CWD: /home/witten/Downloads
sys.argv: ['/home/witten/Downloads/borg-env/bin/borg', 'rinfo', '--json', '--repo', '/home/witten/Downloads/test.borg']
SSH_ORIGINAL_COMMAND: None
```

witten commented:
More very basic testing with borgmatic:

 * [x] interrupting a `borgmatic create` and then re-running it
 * [x] `borgmatic rlist`
 * [x] `borgmatic list`
 * [x] `borgmatic prune`
 * [x] `borgmatic delete`
 * [x] `borgmatic compact`
 * [x] `borgmatic break-lock`
 * [x] `borgmatic key export`

ThomasWaldmann commented:
maybe i can catch that exception and re-raise a borg exception as borg1 did. Von meinem iPhone gesendetAm 01.09.2024 um 17:00 schrieb Dan Helfman ***@***.***>:﻿
Initial feedback... Running the following on a non-existent repo directory (which borgmatic does to probe whether a configured repo already exists prior to rcreate):
$ ~/Downloads/borg-env/bin/borg rinfo --json --repo ~/Downloads/test.borg 
Results in this (exit code 2):
Local Exception

Error:

BackendDoesNotExist: posixfs storage base path does not exist or is not a directory: /home/witten/Downloads/test.borg

If reporting bugs, please include the following:

Traceback (most recent call last):
  File "/home/witten/Downloads/borg/src/borg/archiver/__init__.py", line 622, in main
    exit_code = archiver.run(args)
                ^^^^^^^^^^^^^^^^^^
  File "/home/witten/Downloads/borg/src/borg/archiver/__init__.py", line 516, in run
    rc = func(args)
         ^^^^^^^^^^
  File "/home/witten/Downloads/borg/src/borg/archiver/_common.py", line 151, in wrapper
    with repository:
  File "/home/witten/Downloads/borg/src/borg/repository.py", line 133, in __enter__
    self.open(exclusive=bool(self.exclusive), lock_wait=self.lock_wait, lock=self.do_lock)
  File "/home/witten/Downloads/borg/src/borg/repository.py", line 180, in open
    self.store.open()
  File "/home/witten/Downloads/borg-env/lib/python3.12/site-packages/borgstore/store.py", line 65, in open
    self.backend.open()
  File "/home/witten/Downloads/borg-env/lib/python3.12/site-packages/borgstore/backends/posixfs.py", line 54, in open
    raise BackendDoesNotExist(
borgstore.backends.errors.BackendDoesNotExist: posixfs storage base path does not exist or is not a directory: /home/witten/Downloads/test.borg

Platform: Linux flux 6.1.99-1-MANJARO #1 SMP PREEMPT_DYNAMIC Mon Jul 15 21:35:02 UTC 2024 x86_64
Linux: Unknown Linux  
Borg: 2.0.0b10.dev80+g74a67749  Python: CPython 3.12.4 msgpack: 1.0.8 fuse: None [pyfuse3,llfuse]
PID: 2588631  CWD: /home/witten/Downloads
sys.argv: ['/home/witten/Downloads/borg-env/bin/borg', 'rinfo', '--json', '--repo', '/home/witten/Downloads/test.borg']
SSH_ORIGINAL_COMMAND: None


—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were assigned.Message ID: ***@***.***>

witten commented:
> maybe i can catch that exception and re-raise a borg exception as borg1 did

That'd be great. To be clear, borgmatic does "catch" the Borg error and proceed with the `rcreate`. It's just that the displayed traceback doesn't make for the nicest user experience.

ThomasWaldmann commented:
I am currently working on manifest.archives.

The archives list in borg 1.x used to be contained in the manifest and that had some good and some bad effects:
- borg locked the repo, loaded manifest (and archives list), created an archive, modified manifest (and archives list), saved manifest, unlocked repo
- bad: has an exclusive lock all the time
- bad: even in this PR could not yet get rid of the lock due to that long read-modify-write operation, although storing chunks to the repo does not need the lock anymore.
- good: loaded a "signed list of archives", modified it on the client (e.g. adding or deleting an archive), saved a "signed list of archives"

This PR:
- archives list is `archives/*` directory in the borgstore, each directory entry is a separate object and points to an archive metadata chunk
- good: no problem adding another object there or deleting an object, no r-m-w needed, no lock needed (todo)
- bad: no signed list of archives anymore (but each archive metadata chunk is signed as usual). removing an archive entry is possible on the repo side without the key. but otoh, an attacker with repo r/w access can do arbitrary damage to the repo anyway, including removing it completely.

ThomasWaldmann commented:
Yay, just had multiple borg (all running on same machine / same user) share the same repo at the same time (for borg create, mount, list, ... - excluding borg check and compact). :-)

Only worked when I set `BORG_CACHE_IMPL=adhoc` so it doesn't lock the cache (otherwise the cache lock prevents parallel access). Due to the missing files cache in that mode, it was a bit slow, because all files were `A(dded)`.

ThomasWaldmann commented:
@m3nu did you find something in your experiments with this code?

ThomasWaldmann commented:
Bug or feature?

borg delete and prune now work in a very "quick and dirty" way: they just remove the deleted archive(s) from the `archives/*` directory in the borgstore.

If one runs `borg compact` afterwards, it notices a lot of unused objects and actually deletes them all (this includes the main archive metadata object, the archived items metadata stream and also the items content chunks, if not otherwise used).

Bug? If one runs `borg check --repair` before `borg compact` AND the manifest object is missing or corrupted, it will rebuild the manifest object and also checks for missing entries in the archives directory. It does that by iterating over all objects in the repo and if it finds these main archive metadata objects, it creates an archives directory entry for them. That means it **UNDELETES** previously deleted archives.

Feature? Guess if you accidentally deleted archives and you want them back, this could be seen as a feature.

witten commented:
This may be a pain, but this "feature" might be nice to have—if the user explicitly asks for it. So like a `borg check --repair --undelete` flag or something. I have no idea how feasible that is, but I think it'd be a little unexpected if I'm merely `borg check --repair`ing a broken repository and my deleted archives come back from the dead without me asking for that.

ThomasWaldmann commented:
Rebuilding a missing or corrupted manifest is now much faster, as it won't scan the whole repo for archives anymore.

It only does that scan for `borg check --repair --undelete-archives`.

ThomasWaldmann commented:
Guess 2024-09-09 would be a nice b10 release date.

If somebody finds some major issue, hold me back. :-)

ThomasWaldmann commented:
I'll merge this now. b10 tomorrow and I also might work on some other stuff.

callegar commented:
Some questions about absence of packs/segments.

The restic documentation hints at the importance of packages and on the tunability of the package size where it says (emphasis is mine):
> In certain instances, such as very large repositories (in the TiB range) or very fast upload connections, it is desirable to use larger pack sizes to reduce the number of files in the repository and improve upload performance. **Notable examples are OpenStack Swift and some Google Drive Team accounts, where there are hard limits on the total number of files.**

- With borg moving to an arrangement where the borgstore can be managed by multiple backend and reside on different "providers" (local/cloud) can this become an issue?
- What happens when borg runs out of storage space, but not in terms of storage itself, but of available files?
- Is there provision to bring back packs and if so where would the "packaging" be implemented? In the borgstore backend or "before" the borgstore?
- Would packages let one use a more aggressive (smaller) chunk size? In restic there is a very interesting bug report, where they compare deltas sized by chunk size (with tests based on zpaq).
    - Apparently, the delta size grows rapidly with the chunk size: at 8kB chunks (bup'd default), the deltas are 115% than at 4kB; at 64kB (zpaq's default) they are almost twice as large than at 4kBchunks; at 1MB (restic's default) the deltas are more than 3 times as large than at 4kB. The borg default, at 2MB chunks, makes the deltas 320% the size as with 4kB chunks.
    - Clearly this depends very much on the corpus that is backed up. However, there is a very clear trend: the increase in the delta size slows down for very large chunk sizes, because at a certain point file based deduplication becomes prevalent. Yet, when this happens, chunking gets effectively "switched off" for most files.
    - As the restic developers make clear, the advantage of small chunks largely depends on the efficiency of the indexing approach, because for every chunk you end up paying a price in the indexes which may well hinder the gain on the delta sizes. Specifically, the restic developers at this point clarify that they simply cannot have chunks smaller than 1MB with their current indexing approach. Even the bup's developers say that they might need to move to something somehow larger than 8kB. What is the current situation with the borg2 indexing efficiency?

ThomasWaldmann commented:
@callegar Please read #8572 and ask all remaining questions there.
